cincinnati/src/plugins/internal/arch_filter.rs

```diff
@@ -0,0 +1,343 @@
+//! This plugin can be used to filter a graph by architecture.
+//! It reads the requested architecture from the parameters value at key "arch",
+//! and the value must match the regex specified at ARCH_VALIDATION_REGEX_STR
+//!
+//! The filtering also removes any architecture suffixes from the version strings
+//! if they are present. The assumption for this is that the architecture would
+//! be encoded as part of the _build_ information according to the SemVer specification.
+use crate::plugins::{
+    AsyncIO, BoxedPlugin, InternalIO, InternalPlugin, InternalPluginWrapper, PluginSettings,
+};
+use commons::GraphError;
+use failure::{Fallible, ResultExt};
+use futures::Future;
+use prometheus::Registry;
+
+pub static DEFAULT_KEY_FILTER: &str = "io.openshift.upgrades.graph";
+pub static DEFAULT_ARCH_KEY: &str = "release.arch";
+pub static DEFAULT_DEFAULT_ARCH: &str = "amd64";
+pub static DEFAULT_DEFAULT_ARCH_THRESHOLD_VERSION: &str = "4.2.0-rc.0";
+
+#[derive(Clone, Debug, Deserialize, SmartDefault)]
+#[serde(default)]
+pub struct ArchFilterPlugin {
+    #[default(DEFAULT_KEY_FILTER.to_string())]
+    pub key_prefix: String,
+
+    #[default(DEFAULT_ARCH_KEY.to_string())]
+    pub key_suffix: String,
+
+    #[default(DEFAULT_DEFAULT_ARCH.to_string())]
+    pub default_arch: String,
+
+    #[default(DEFAULT_DEFAULT_ARCH_THRESHOLD_VERSION.to_string())]
+    pub default_arch_threshold_version: String,
+}
+
+impl PluginSettings for ArchFilterPlugin {
+    fn build_plugin(&self, _: Option<&Registry>) -> Fallible<BoxedPlugin> {
+        Ok(new_plugin!(InternalPluginWrapper(self.clone())))
+    }
+}
+
+impl ArchFilterPlugin {
+    pub const PLUGIN_NAME: &'static str = "arch-filter";
+
+    /// Validate plugin configuration and fill in defaults.
+    pub fn deserialize_config(cfg: toml::Value) -> Fallible<Box<dyn PluginSettings>> {
+        let plugin: Self = cfg.try_into()?;
+
+        ensure!(!plugin.key_prefix.is_empty(), "empty arch-key prefix");
+        ensure!(!plugin.key_suffix.is_empty(), "empty arch-key suffix");
+
+        Ok(Box::new(plugin))
+    }
+}
+
+/// Evaluate an architecture from the combination of "arch" and "version" parameters.
+///
+/// Assuming that version has been sent since the beginning of all versions,
+/// this function fails if `arch.is_none() && version > default_arch_threshold_version`.
+fn infer_version(
```

Comment:
nit: `infer_arch_from_version`?

Reply:
I didn't want to over-specify, as we might include other data -- e.g. telemetry data from the clusterID -- later on.

Reply:
Coming back to this, I failed to read this properly. I almost agree with your nit. My intention was to name it `infer_arch` and not specify the information considered for the inference in the function name.